### PR TITLE
Add REST API

### PR DESCRIPTION
**Pull request by drewwalters96**
- Add Dockerfile
- Add requirements.txt
- Add server.py
- Add template rendering

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,6 @@
+FROM python:2.7-slim
```
Comment (smaeul) [inline Dockerfile]: Ain't no way we use python 2

File: server.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
+from flask import Flask
+import json
+
+app = Flask(__name__)
+
+
+@app.route('/status')
```
Comment (smaeul) [inline server.py]: Please sort endpoints

**Review by smaeul — DISMISSED** [review]

File: server.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+@app.route('/')
+def index(): pass
+
+
+@app.route('/login')
+def login(): pass
+
+
+@app.route('/user/<username>')
```
Comment (smaeul) [inline server.py]: The username might not be safe to put in a URL, so we should use the numeric ID here

File: server.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+@app.route('/book/<b>')
+def show_book(b):
+    """ Returns a page with information about the book, maybe list of classes
+    it is used in? Below that is a few nearby postings with a button to see
+    all, which takes you to /postings?book=b """
+    return render_template('book.html', book=b)
+
+
+@app.route('/book/<b>/posting/<p>')
```
Comment (drewwalters96) [inline server.py]: Why is this route dependent on book?

File: server.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+def show_book_posting(b, p):
+    """ Shows a single posting, with a map, pics, description, and a link to
+    send the seller a message """
+    return render_template('posting.html', book=b, posting=p)
+
+
+@app.route('/book/<b>/postings')
```
Comment (drewwalters96) [inline server.py]: Same question as above

File: server.py [hunk 6]
```diff
@@ -66,6 +93,19 @@ def api_user(u):
     return json.dumps({})
 
 
+@app.route('/api/user/<u>/messages')
+def api_user_messages(u):
+    """ Gets an array of messages sent or received by a user """
+    return json.dumps([])
+
+
+@app.route('/api/user/<u>/messages/new')
```
Comment (drewwalters96) [inline server.py]: Only `GET` is allowed by default so the decorator will need `@app.route('/api/user/<u>/messages/new', methods=['GET', 'POST'])`

File: server.py [hunk 7]
```diff
@@ -31,14 +31,29 @@ def api_book_postings(b):
     return json.dumps([])
 
 
+@app.route('/api/book/<b>/postings/new')
```
Comment (drewwalters96) [inline server.py]: Only GET is allowed by default so the decorator will need @app.route('/api/user/<u>/messages/new', methods=['GET', 'POST'])

Comment (drewwalters96) [inline server.py]: It might be a good idea to add version numbers to each endpoint for backwards compatibility i.e. `/api/v1`

File: server.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
+from flask import Flask, render_template
```
Comment (drewwalters96) [inline server.py]: Please read [PEP257](https://www.python.org/dev/peps/pep-0257/) which outlines docstring conventions

File: server.py (hunk 3 above)
Comment (drewwalters96) [reply server.py]: Why is this unsafe if the username will be visible on the website? The API should be able to filter by public username

**Review by smaeul — APPROVED** [review]

**Pull request merged**